x/firewood/sync_test.go

```diff
@@ -0,0 +1,274 @@
+)
+
+var (
+	rangeProofMarshaler  = RangeProofMarshaler{}
+	changeProofMarshaler = ChangeProofMarshaler{}
+)
+
+// Test_Firewood_Sync tests syncing a Firewood database from empty to a full database.
+func Test_Firewood_Sync(t *testing.T) {
+	t.Parallel()
+	dbSizes := []int{0, 1, 1_000, 10_000, 100_000}
+	for _, serverSize := range dbSizes {
+		t.Run(fmt.Sprintf("numKeys=%d", serverSize), func(t *testing.T) {
+			t.Parallel()
+			testSync(t, 0, serverSize)
+		})
+	}
+}
+
+// Test_Firewood_Sync_Overwrite tests syncing when the client already has some data.
+// This data is expected to be overwritten by the server's data.
+func Test_Firewood_Sync_Overwrite(t *testing.T) {
+	t.Parallel()
+	dbSizes := []int{0, 1, 1_000, 10_000, 100_000}
+	initialSize := 500
+	for _, serverSize := range dbSizes {
+		t.Run(fmt.Sprintf("numKeys=%d", serverSize), func(t *testing.T) {
+			t.Parallel()
+			testSync(t, initialSize, serverSize)
+		})
+	}
+}
+
+func testSync(t *testing.T, clientKeys int, serverKeys int) {
+	require := require.New(t)
+	seed := time.Now().UnixNano()
+	t.Log("seed:", seed)
+	fullDB := generateDB(t, serverKeys, seed)
+	db := generateDB(t, clientKeys, seed+1) // guarantee different data
+
+	ctx := t.Context()
+	root, err := fullDB.GetMerkleRoot(ctx)
+	require.NoError(err)
+
+	syncer, err := xsync.NewSyncer(
+		db,
+		xsync.Config[*RangeProof, *ChangeProof]{
+			RangeProofMarshaler:   rangeProofMarshaler,
+			ChangeProofMarshaler:  changeProofMarshaler,
+			RangeProofClient:      p2ptest.NewSelfClient(t, ctx, ids.EmptyNodeID, xsync.NewGetRangeProofHandler(fullDB, rangeProofMarshaler)),
+			ChangeProofClient:     p2ptest.NewSelfClient(t, ctx, ids.EmptyNodeID, xsync.NewGetChangeProofHandler(fullDB, rangeProofMarshaler, changeProofMarshaler)),
+			SimultaneousWorkLimit: 5,
+			Log:                   logging.NoLog{},
+			TargetRoot:            root,
+			EmptyRoot:             ids.ID(types.EmptyRootHash),
+		},
+		prometheus.NewRegistry(),
+	)
+	require.NoError(err)
+	require.NotNil(syncer)
+
+	require.NoError(syncer.Start(ctx))
+	err = syncer.Wait(ctx)
+	if errors.Is(err, xsync.ErrFinishedWithUnexpectedRoot) {
+		t.Log("syncer reported root mismatch; logging diff between DBs")
+		logDiff(t, fullDB, db)
+	}
+	require.NoError(err)
+}
+
+// generateDB creates a new Firewood database with up to [numKeys] random key/value pairs.
+// The database will be automatically closed when the test ends.
+// Note that each key/value pair may not be unique, so the resulting database may have fewer than [numKeys] entries.
+func generateDB(t *testing.T, numKeys int, seed int64) *syncDB {
+	t.Helper()
+	path := filepath.Join(t.TempDir(), "firewood.db")
+
+	fw, err := ffi.New(path, ffi.DefaultConfig())
+	require.NoError(t, err)
+	require.NotNil(t, fw)
+	t.Cleanup(func() {
+		ctx := context.WithoutCancel(t.Context())
+		ctx, cancel := context.WithTimeout(ctx, 5*time.Second) // allow some time for garbage collection
+		defer cancel()
```

Comment:
Why do we do this?

Reply:
> It is technically racy, but I would be really, really surprised if it ever flaked.

It being racey is a reason to not include it - if a unit test is flaky is is not a useful signal and especially on stressed CI workers sleeps being used to patch racy logic is unreliable. 

> What would your alternative be? There's no other decent context to pass to this function, and any subsequent test runs are unaffected